File: threshold_frost/src/commitment.go
package threshold_signing

import (
	"crypto/rand"
	"io"

	thres "github.com/ArkLabsHQ/thresholdmagic/thresholdcore"

	"github.com/decred/dcrd/dcrec/secp256k1/v4"
)

type SigningNonce struct {
	hiding  *secp256k1.ModNScalar
	binding *secp256k1.ModNScalar

	commitments SigningCommitments
}

func NewNonce(secret thres.SecretShare, rng io.Reader) SigningNonce {
	hiding := generateNonce(secret, rng)
	binding := generateNonce(secret, rng)

	hidingCommitment := secp256k1.JacobianPoint{}
	bindingCommitment := secp256k1.JacobianPoint{}

	secp256k1.ScalarBaseMultNonConst(&hiding, &hidingCommitment)
	secp256k1.ScalarBaseMultNonConst(&binding, &bindingCommitment)

	commitments := SigningCommitments{
		Hiding:  hidingCommitment,
		Binding: bindingCommitment,
	}

	return SigningNonce{
		hiding:      &hiding,
		binding:     &binding,
		commitments: commitments,
	}
}

type SigningCommitments struct {
	Binding secp256k1.JacobianPoint
	Hiding  secp256k1.JacobianPoint
}

type GroupCommitmentShare struct {
	elem secp256k1.JacobianPoint
}

func (s *SigningCommitments) toGroupCommitmentShare(bindingScalar secp256k1.ModNScalar) GroupCommitmentShare {
	// sum = B_i + b_i * H_i
	var bH secp256k1.JacobianPoint

	secp256k1.ScalarMultNonConst(&bindingScalar, &s.Binding, &bH)
	sum := secp256k1.JacobianPoint{}
	secp256k1.AddNonConst(&s.Hiding, &bH, &sum)
	return GroupCommitmentShare{elem: sum}
}

func generateNonce(secret thres.SecretShare, rng io.Reader) secp256k1.ModNScalar {

	var rb [32]byte
	if rng == nil {
		rng = rand.Reader
	}
	_, _ = io.ReadFull(rng, rb[:])

	secretBytes := (*secp256k1.ModNScalar)(&secret).Bytes()

	concatenatedBytes := append(rb[:], secretBytes[:]...)

	return H3(concatenatedBytes[:])

}

type GroupCommitment struct {
	elem secp256k1.JacobianPoint
}

func ComputeGroupCommitment(
	s *SigningPackage,
	bfl BindingFactorList,
) (GroupCommitment, error) {
	identity := secp256k1.JacobianPoint{}
	groupCommitment := secp256k1.JacobianPoint{}

	n := len(s.Commitments)

	bindingScalars := make([]secp256k1.ModNScalar, 0, n)
	bindingElements := make([]secp256k1.JacobianPoint, 0, n)

	for id, comm := range s.Commitments {
		bind := comm.Binding
		hide := comm.Hiding

		// Prevent identity commitments
		if identity.EquivalentNonConst(&bind) || identity.EquivalentNonConst(&hide) {
			return GroupCommitment{}, ErrIdentityCommitment
		}

		// lookup binding factor for id
		bf, ok := bfl.Get(id)
		if !ok {
			return GroupCommitment{}, ErrUnknownIdentifier
		}

		// collect for single MSM
		bindingElements = append(bindingElements, bind)
		bindingScalars = append(bindingScalars, bf.Scalar)

		// sum hiding commitments
		temp := secp256k1.JacobianPoint{}
		secp256k1.AddNonConst(&groupCommitment, &hide, &temp)
		groupCommitment = temp
		//
	}

	// accumulated binding commitment via MSM
	acc := VartimeMultiscalarMul(bindingScalars, bindingElements)
	temp := secp256k1.JacobianPoint{}
	secp256k1.AddNonConst(&groupCommitment, acc, &temp)
	groupCommitment = temp

	return GroupCommitment{elem: groupCommitment}, nil
}

File: dart/threshold/lib/frost/commitment.dart
import 'dart:math';
import 'dart:typed_data';

import 'package:pointycastle/ecc/api.dart';
import 'package:threshold/core/identifier.dart';
import 'package:threshold/core/share.dart'; // for SecretShare
import 'package:threshold/core/utils.dart';
import 'package:threshold/frost/errors.dart';
import 'package:threshold/frost/hasher.dart';
import 'package:threshold/frost/utils.dart';
import 'package:threshold/frost/binding.dart';

class SigningNonce {
  final BigInt hiding;
  final BigInt binding;
  final SigningCommitments commitments;

  SigningNonce(this.hiding, this.binding, this.commitments);
}

class SigningCommitments {
  final ECPoint binding;
  final ECPoint hiding;

  SigningCommitments(this.binding, this.hiding);
}

class GroupCommitmentShare {
  final ECPoint elem;
  GroupCommitmentShare(this.elem);
}

class GroupCommitment {
  final ECPoint elem;
  GroupCommitment(this.elem);
}

// Generates a new nonce pair and returns the SigningNonce struct
SigningNonce newNonce(SecretShare secret) {
  final hiding = generateFrostNonce(secret);
  final binding = generateFrostNonce(secret);

  final hidingCommitment = (secp256k1Curve.G * hiding)!;
  final bindingCommitment = (secp256k1Curve.G * binding)!;

  final commitments = SigningCommitments(bindingCommitment, hidingCommitment);

  return SigningNonce(hiding, binding, commitments);
}

BigInt generateFrostNonce(SecretShare secret) {
  final rand = Random.secure();
  final rb = Uint8List(32);
  for (var i = 0; i < 32; i++) {
    rb[i] = rand.nextInt(256);
  }

  // secret is BigInt (typedef SecretShare = BigInt)
  final secretBytes = bigIntToBytes(secret);
  final concatenatedBytes = Uint8List.fromList(rb + secretBytes);

  return h3(concatenatedBytes);
}

extension SigningCommitmentsExt on SigningCommitments {
  GroupCommitmentShare toGroupCommitmentShare(BigInt bindingScalar) {
    // sum = B_i + b_i * H_i (NOTE: Go code logic: sum = Hiding + bindingScalar * Binding)
    // Wait, looking at Go code:
    // secp256k1.ScalarMultNonConst(&bindingScalar, &s.Binding, &bH)
    // secp256k1.AddNonConst(&s.Hiding, &bH, &sum)
    // So it is Hiding + scalar * Binding

    final bH = (binding * bindingScalar)!;
    final sum = (hiding + bH)!;
    return GroupCommitmentShare(sum);
  }
}

class SigningPackage {
  final Map<Identifier, SigningCommitments> commitments;
  final Uint8List message;

  SigningPackage(this.commitments, this.message);

  SigningCommitments? signingCommitment(Identifier id) {
    return commitments[id];
  }
}

// computeGroupCommitment
GroupCommitment computeGroupCommitment(
  SigningPackage s,
  BindingFactorList bfl,
) {
  var groupCommitment = secp256k1Curve.curve.infinity!;

  final bindingScalars = <BigInt>[];
  final bindingElements = <ECPoint>[];

  for (final entry in s.commitments.entries) {
    final id = entry.key;
    final comm = entry.value;
    final bind = comm.binding;
    final hide = comm.hiding;

    if (bind.isInfinity || hide.isInfinity) {
      throw errIdentityCommitment;
    }

    // lookup binding factor
    final bf = bfl.get(id);
    if (bf == null) {
      throw errUnknownIdentifier;
    }

    bindingElements.add(bind);
    bindingScalars.add(bf.scalar);

    // sum hiding commitments
    groupCommitment = (groupCommitment + hide)!;
  }

  // accumulated binding
  final acc = vartimeMultiscalarMul(bindingScalars, bindingElements);
  groupCommitment = (groupCommitment + acc)!;

  return GroupCommitment(groupCommitment);
}
